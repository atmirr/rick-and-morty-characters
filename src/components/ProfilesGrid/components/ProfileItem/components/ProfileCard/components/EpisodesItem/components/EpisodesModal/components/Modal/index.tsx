import React, { useEffect, useState, useCallback } from 'react';
import styled, { css, keyframes } from 'styled-components';
import { createPortal } from 'react-dom';

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

const Container = styled.div`
  position: fixed;
  overflow: hidden;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.65);
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 1.5em 1em;
  z-index: 999999;
  box-sizing: border-box;
  cursor: pointer;
  ${({ fade }: { fade: boolean }) =>
    fade &&
    css`
      animation: ${fadeIn} 0.3s 1 linear;
      animation-fill-mode: forwards;
      opacity: 0;
    `}
`;

const Overlay = styled.div`
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
`;

type PropTypes = {
  children: React.ReactNode;
  fade: boolean;
  defaultOpened: boolean;
  handleCloseModal: () => void;
};

export function Modal({
  children,
  fade = false,
  defaultOpened = false,
  handleCloseModal,
}: PropTypes): React.ReactElement {
  const documentBody = document.getElementsByTagName('body')[0];
  const [isOpen, setIsOpen] = useState(defaultOpened);
  const close = useCallback(() => {
    // Making the vertical scrollbar's overflow back to normal when the component is unmounted
    documentBody.style.overflowY = 'auto';
    setIsOpen(false);
    handleCloseModal && handleCloseModal();
  }, []);

  const handleEscape = useCallback(
    (event) => {
      if (event.keyCode === 27) close();
    },
    [close],
  );

  useEffect(() => {
    if (isOpen) {
      // Making the vertical scrollbar lock when the backdrop is mounted
      documentBody.style.overflowY = 'hidden';
      // Add event listener for the Esc key
      document.addEventListener('keydown', handleEscape, false);
    }
    return () => {
      document.removeEventListener('keydown', handleEscape, false);
    };
  }, [handleEscape, isOpen]);

  return createPortal(
    isOpen ? (
      <Container data-testid="modal" fade={fade}>
        <Overlay onClick={close} />
        {children}
      </Container>
    ) : null,
    documentBody,
  );
}

export default Modal;
