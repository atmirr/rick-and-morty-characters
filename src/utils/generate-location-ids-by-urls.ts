export const generateLocationIdsByUrls = (urls: string[]): string[] =>
  urls.reduce((result: string[], url: string) => {
    const splittedUrl = url.split('/');
    const id = splittedUrl[splittedUrl.length - 1];
    return id ? [...result, id] : [...result];
  }, []);
