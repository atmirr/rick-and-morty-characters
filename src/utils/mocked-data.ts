export const characters = [
  {
    id: 1,
    name: 'Rick Sanchez',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:48:46.250Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/1',
  },
  {
    id: 2,
    name: 'Morty Smith',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:50:21.651Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/2',
  },
];

export const episodes = [
  'Rixty Minutes',
  'Something Ricked This Way Comes',
  'Close Rick-counters of the Rick Kind',
  'Ricksy Business',
  'A Rickle in Time',
];

export const locations = [
  {
    name: 'Citadel of Ricks',
    type: 'Space station',
    dimension: 'unknown',
  },
  {
    name: 'Earth (Replacement Dimension)',
    type: 'Planet',
    dimension: 'Replacement Dimension',
  },
];
