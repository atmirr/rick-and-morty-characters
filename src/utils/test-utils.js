import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockedInitialState = {
  characters: { items: [], nextPage: null, error: null, isLoading: false },
  episodes: { items: [], nextPage: null, error: null, isLoading: false },
  locations: { items: [], error: null, isLoading: false },
};

function render(
  ui,
  {
    initialState = mockedInitialState,
    store = configureStore()(initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
