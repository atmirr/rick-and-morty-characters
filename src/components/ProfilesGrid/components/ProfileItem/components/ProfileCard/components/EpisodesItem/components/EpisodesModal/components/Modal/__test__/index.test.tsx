import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, render } from 'src/utils/test-utils';
import { episodes } from 'src/utils/mocked-data';
import Modal from '../index';

const children = (
  <ul>
    {episodes.map((episode, i) => (
      <li key={i}>{episode}</li>
    ))}
  </ul>
);

describe('<Modal />', () => {
  it('should renders the component', () => {
    const { container } = render(<Modal />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not display the modal by default', () => {
    render(<Modal />);
    expect(screen.queryAllByTestId('modal').length).toBe(0);
  });

  it('should display the modal when defaultOpened prop set to true', () => {
    render(<Modal defaultOpened />);
    expect(screen.queryAllByTestId('modal').length).toBe(1);
  });

  it('should display the given children component in the modal when children  has been set', () => {
    const { queryAllByTestId, unmount } = render(
      <Modal defaultOpened>{children}</Modal>,
    );
    // We are using React Portal for the Modal component
    // to attach it to the document.body
    // so we have to consider document.body as our container for our tests
    const container = document.body;
    expect(queryAllByTestId('modal').length).toBe(1);
    expect(container.querySelectorAll('ul').length).toBe(1);
    expect(container.querySelectorAll('li').length).toBe(episodes.length);
    episodes.forEach((episode) => {
      expect(screen.getAllByText(episode).length).toBe(1);
    });
    // Clean up the document.body
    unmount();
  });
});
