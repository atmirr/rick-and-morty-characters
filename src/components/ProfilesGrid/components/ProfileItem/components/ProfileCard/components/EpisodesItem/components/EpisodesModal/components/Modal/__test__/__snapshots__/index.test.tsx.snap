// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should renders the component 1`] = `null`;
